perf-tool/src/exception.cpp

```diff
@@ -72,53 +71,62 @@ JNIEXPORT void JNICALL Exception(jvmtiEnv *jvmtiEnv,
     jvmtiLineNumberEntry *lineTable;
     char buffer[20];
     char *methodName;
+    int numExceptions;
 
-    static int numExceptions = 0; // number of total exceptions recorded
-    numExceptions++;
+    /* Get number of exceptions recorded and increment */
+    numExceptions = atomic_fetch_add(&exceptionSampleCount, 1);
+    jdata["numExceptions"] = numExceptions;
 
-    // First get exception address
-    char ptrStr[] = "%p";
-    sprintf(buffer, ptrStr, exception);
-    string exceptionStr(buffer);
-    jdata["exceptionAddress"] = (char*)exceptionStr.c_str();
+    /* Get exception address */
+    sprintf(buffer, "%p", exception);
+    jdata["exceptionAddress"] = buffer;
 
     /* Get method name */
     err = jvmtiEnv->GetMethodName(method, &methodName, NULL, NULL);
-    jdata["callingMethod"] = (char *)methodName;  // record calling method
+    if (err == JVMTI_ERROR_NONE) {
+        jdata["callingMethod"] = methodName;  // record calling method
+    }
+    err = jvmtiEnv->Deallocate((unsigned char*)methodName);
+    check_jvmti_error(jvmtiEnv, err, "Unable to deallocate methodName");
 
     /* Get line number */
-    err = jvmtiEnv->GetLineNumberTable(method, &lineCount, &lineTable);
+    err = jvmtiEnv->GetLineNumberTable(method, &lineCount, &lineTable);  // returns table of source line num entries
     if (err == JVMTI_ERROR_NONE) { // Find line
-        lineNumber = lineTable[0].line_number;
+        lineNumber = lineTable[0].line_number;  // initially set line number to first line number of lineTable
         for (int i = 1; i < lineCount; i++) {
+            // iterates through lineTable until the location from where the calling method was called is reached
             if (location < lineTable[i].start_location) {
+                // location of line in calling method code has been passed, exit for loop
                 break;
             } else {
+                // location not passed yet; update line number
                 lineNumber = lineTable[i].line_number;
             }
         }
         jdata["callingMethodLineNumber"] = lineNumber;  // record line number of calling method
     }
+    err = jvmtiEnv->Deallocate((unsigned char*)lineTable);
```

Comment:
A similar issue here: I think that if `GetLineNumberTable` fails, there will be no `lineTable` being allocated (to be checked) and therefore we don't have to issue `Deallocate` followed by a crash in `check_jvmti_error`